src/types.ts

```diff
@@ -0,0 +1,221 @@
+  setConfig: <K extends keyof TSchema>(key: K, value: ResolvedConfigValue<TSchema[K]>) => void;
+  getAllConfig: () => TConfig;
+  getWindowValue: <K extends keyof TSchema>(key: K) => ResolvedConfigValue<TSchema[K]> | null;
+  getStorageValue: <K extends keyof TSchema>(key: K) => ResolvedConfigValue<TSchema[K]> | null;
+}
+
+// useAdminConfig types
+export type AdminField<TSchema extends Record<string, Config>, TKey extends keyof TSchema> = TSchema[TKey] & {
+  /**
+   * Schema key of the config
+   */
+  key: TKey;
+  /**
+   * Full path of the config (with `namespace`)
+   */
+  path: string;
+  /**
+   * Value stored in `window.{path}`
+   */
+  windowValue: ResolvedConfigValue<TSchema[TKey]> | null;
+  /**
+   * Value stored in `storage.getItem({path})`
+   */
+  storageValue: ResolvedConfigValue<TSchema[TKey]> | null;
+  /**
+   * True if a value is store on the localstorage
+   */
+  isFromStorage: boolean;
+  /**
+   * Computed value from storage, window, schema[key].default
+   */
+  value: ResolvedConfigValue<TSchema[TKey]>;
+  /**
+   * Value setter
+   */
+  set: (value: ResolvedConfigValue<TSchema[TKey]>) => void;
+};
+
+type Lookup<T, K> = K extends keyof T ? T[K] : never;
+type TupleFromInterface<T, K extends Array<keyof T> = Array<keyof T>> = {
+  [I in keyof K]: Lookup<T, K[I]>;
+};
+
+export type AdminFields<TSchema extends Record<string, Config>> = TupleFromInterface<
+  {
+    [key in keyof TSchema]: AdminField<TSchema, key>;
+  }
+>;
+
+// useAdminConfig generic types
+type AdminProps<T, U = T> = {
+  key: string;
+  path: string;
+  windowValue: T | null;
+  storageValue: T | null;
+  isFromStorage: boolean;
+  value: T;
+  set: (value: U) => void;
+};
+
+/**
+ * `useAdminConfig.fields` in a generic version.
+ *
+ * This should be used if you are implement a generic component
+ * that consume any `fields` as prop.
+ *
+ * Note: "custom" type and "string" with enum are defined as `any` to be
+ * compatible with any schemas. You will need to validate them in your
+ * implementation to retrieve type safety.
+ */
+export type GenericAdminFields = Array<
+  | (StringConfig & AdminProps<string>)
+  | (NumberConfig & AdminProps<number>)
+  | (BooleanConfig & AdminProps<boolean>)
+  | (StringEnumConfig & AdminProps<string, any>)
+  | (CustomConfig<any> & AdminProps<any>)
+>;
+
+// useConfig types
+type UseConfigReturnType<TSchema extends Record<string, Config>> = {
+  getConfig: <K extends keyof TSchema>(path: K) => ResolvedConfigValue<TSchema[K]>;
+  getAllConfig: () => ResolvedSchema<TSchema>;
+  setConfig: <K extends keyof TSchema>(path: K, value: ResolvedConfigValue<TSchema[K]>) => void;
+};
+
+type Namespaced<T, TNamespace extends string> = {
+  [P in keyof T as P extends `getConfig`
+    ? `get${Capitalize<TNamespace>}Config`
+    : P extends "getAllConfig"
+    ? `getAll${Capitalize<TNamespace>}Config`
+    : `set${Capitalize<TNamespace>}Config`]: T[P];
+};
```

Comment:
I have hard time understanding what is happening here. And tripple ternary does not make it easier :)
An asdoc description could be helpful